### DISCUSSION TITLE
Setting up Sharing - DDNS through cPanel?

### DISCUSSION THREAD

**Discussion by oldgristmill** [mcguirepr89/BirdNET-Pi#172] [Q&A]
Trying to get my BirdNET-Pi accessible, and noticed that cPanel has its own DDNS service built in:

https://docs.cpanel.net/cpanel/domains/dynamic-dns/

Looks like I can just set a cron job to make the appropriate webcall from the pi for now, but for end-user simplicity it would be nice to bake that option into the settings UI or at least into the birdnet.conf.

**Comment by oldgristmill** [comment]
So yes, that worked. Almost.

Created a new subdomain in cPanel: https://birds.oldgristmill.ca and then created a cPanel DDNS entry. Copied the generated webcall address into a cron job that runs every five mins on the pi and after the first time it's clear it's working. So now the subdomain is connected to the public IP address of my router.

The router is set up to forward port 80 and 443 to the LAN IP reserved for the pi. Also updated the BirdNET-Pi URL in "Advanced Settings" to https://birds.oldgristmill.ca.

The A record shows up on DNS lookup tools and it all seems fine.

But visiting the URL above doesn't connect with the Pi. Troubleshooting suggestions?

**Comment by mcguirepr89** [comment]
Hi, @oldgristmill 

I'm not sure I understand the relationship between your BirdNET-Pi installation and your cPanel installation, but I can certainly provide troubleshooting suggestions and can answer questions regarding setting up DDNS for the BirdNET-Pi through https://noip.com.

Also, this may be better suited as an issue?

To be thorough, can you confirm the following:
- Ports 80 and 443 are forwarded to the BirdNET-Pi
- The A record for birds.oldgristmill.ca should be pointing to 154.5.93.219

and can you let me know just for clarification which branch you're using (since some folks are testing `newif` and some are on `main`).

Basic troubleshooting for something like this:
1. Take a look at `/home/pi/BirdNET-Pi/birdnet.conf` to double-check that "`BIRDNETPI_URL`" is set 
1. Take a look at the `/etc/caddy/Caddyfile` to ensure https://birds.oldgristmill.ce is listed in the site block
1. Try reloading `caddy` . . .
    ```
    sudo systemctl reload caddy
    ```
    then look at the log
    ```
    journalctl -eu caddy
    ```

If the issue is on the BirdNET-Pi, it will be in the `Caddyfile` -- if there is an issue in the `Caddyfile`, the log will show it.

That will _almost_ definitely tell you what is going wrong.

If that does _not_ show you what is going on, and you still suspect the BirdNET-Pi as the culprit, do the following:
1. Issue `stop_core_services.sh`
1. Try restarting `caddy`
    ```
    sudo systemctl restart caddy
    ```
1. Then tail the `/var/log/syslog` for something _seriously_ wrong with the pi. If you don't see anything super obvious, then it is not the BirdNET-Pi, and is related to the configuration of your LAN, likely your routing table or firewall.

Always happy to review logs, just let me know what you see and I'll let you know what I think!

My best,
Patrick

**Reply by oldgristmill** [reply]
Tried to restart caddy and it timed out; the log contents are attached. When I do a port scan, port 80 on my public IP reports that it's open, but port 443 reports that it's closed. Port forwarding seems correct on my router, so it seems the pi is refusing the https connection?

[gristmill-caddylog.txt](https://github.com/mcguirepr89/BirdNET-Pi/files/8347388/gristmill-caddylog.txt)

**Reply by oldgristmill** [reply]
I'm on the main branch for now. 

As for the relationship between cPanel and BirdNet-Pi, I just discovered that cPanel (a site and server management platform popular with web hosting companies, and used by my own webhost) has DDNS features built in, and because of how it's directly connected to the management of your hosted website, it takes very little effort to set up and use with BirdNet-Pi (in principle). 

Once I get my own instance working, I'd be happy to provide documentation for it for anyone else who might benefit in the future. Right now, it seems the problem lies either with my port forwarding set up on my router or with the Pi itself.

**Reply by mcguirepr89** [reply]
Hi, @oldgristmill 

Caddy timed out when attempting to handle ACME domain ownership challenges because no CA (LetsEncrypt) could reach port 443. This means that the issue lives on the network and its handling of requests to port 443. Caddy is not getting any responses from CAs (which will only reply on 443 to complete the SSL challenge).

I can confirm that this is the issue from the port scanning on my end:
```
patrick @ mcchrome ➜  ~  sudo nmap -p443 154.5.93.219                 

Starting Nmap 7.80 ( https://nmap.org ) at 2022-03-25 09:38 EDT
Nmap scan report for d154-5-93-219.bchsia.telus.net (154.5.93.219)
Host is up (0.11s latency).

PORT    STATE  SERVICE
443/tcp closed https

Nmap done: 1 IP address (1 host up) scanned in 0.52 seconds
```

So you'll want to keep tinkering with your router settings to ensure that the traffic gets forwarded as needed.

Let me know if you'd like to do a remote desktop session via AnyDesk and I'll be happy to poke around and see what I can find.

My best!

**Reply by oldgristmill** [reply]
So it seems my ISP (Telus, here in Canada) blocks incoming connections to port 443 for "security" reasons. That's a frustrating problem I'm not sure how to work around... Thoughts?

**Reply by mcguirepr89** [reply]
Hmmm . . . I have thought about this a good bit and I can only think of three basic options right now. The list below is in an order based on what I assume your needs will be when trying to make this publicly available.

1. Try obtaining a Wild-Card SSL for *.oldgristmill.ca (which would be able to cover your birds.oldgristmill.ca hostname) and manually install it as a usable certificate for Caddy. At that point, you could reconfigure `caddy` to serve all HTTPS traffic over a trivial port that your ISP isn't blocking (8443 perhaps?). This would allow you the proper TLS without needing port 443 to participate. The down side to this is that your BirdNET-Pi URL will require a port specification, e.g., https://birds.oldgristmill.ca:8443. You may, however, be funneling visitors to the site via links, in which case adding the port designation to the URL link isn't too cumbersome nor difficult  and the user wouldn't really notice.

2. Configure Caddy to issue its own SSL using whatever trivial port you find your ISP isn't blocking. The down side here is that your visitors will be warned in an unfriendly way that no verified Certificate Authority has signed the certificate and that the ownership of the site cannot be verified. A lot of users will get to freaked out to proceed and will probably not really use the site.

3. Serve everything over HTTP and discontinue any and all sensitive data transmission (i.e., no longer utilize the web-terminal, web-based settings interface, live audio stream)

I will keep scratching my head over this to see if I can think of another way to work around this ISP ding-dongery.

My best regards from Virginia,
Patrick

**Reply by oldgristmill** [reply]
I'd like to try for #3 for now --I can always access the pi locally for settings, terminal, etc; but maybe in the future I could figure out how to have the DB, recordings, spectrographs, etc replicated offsite at my webhost and provide access to those, instead. 

In the meantime, how can I adjust files/settings to just serve the unsecured stuff so that you all can admire the overwhelming numbers of waxwings and robins drowning out everything else right now? :D 

(Seriously, I need to move the microphone; it's close to a rowan/mountain ash that those two species are chowing down right now, making it hard to hear any of the other species around right now...)

**Reply by oldgristmill** [reply]
I think I managed to get a certificate for the subdomain the Pi will use... now to figure out how to get it onto the pi, configure caddy as appropriate and go from there. I guess I need to search for some Caddy documentation...

**Reply by mcguirepr89** [reply]
https://www.caddyserver.com is _the_ resource for Caddy -- [start with this](https://caddyserver.com/docs/caddyfile) just to get an idea of how to read the Caddyfile. and [this page specifically](https://caddyserver.com/docs/caddyfile/directives/tls#tls) will help you out on what to do with your certificate and key. You want the very first example among the examples at the bottom.

I've not had the need to do this myself, so my guidance is just based on understanding, but it should point you in the right direction and get you started.

You should be able to easily view your Caddyfile from the "System" > "Info" link. That viewer chops it up and makes it look ugly and doesn't show indentations, but the line breaks are accurate. It'll be good to just have an idea of that file while you proceed.

I'm embarrassed to say I can't remember exactly everything what you'll see for your installation since I've been working with the development branch `newif` for the last few weeks. So, if something I mention is vague, it's not intentional and the specifics are available.

--
To get the certificate and key file onto the pi, you should be able to use the web interface, "System" > "Tools" > "File Manager" and upload the certificates as files somewhere, then use the web terminal, (or SSH) to issue something along these lines:
Note: I'm assuming you will upload the certificate and key files to the `scripts` directory from the "File Manager" in this example

- Make a directory for the files where Caddy likes them:
   ```
   sudo mkdir -p /var/lib/caddy/.local/share/caddy/certificates/mycerts/birds.oldgristmill.ca
   ```
- Move the certificates to the new directory:
   ```
   sudo mv /home/pi/BirdNET-Pi/scripts/yourcertificate.pem /home/pi/BirdNET-Pi/scripts/yourkey.pem /var/lib/caddy/.local/share/caddy/certificates/mycerts/birds.oldgristmill.ca
   ```
- Change the directory and file ownership and permissions:
   ```
   sudo chown -R caddy:caddy /var/lib/caddy/.local/share/caddy/certificates/mycerts
   sudo chmod -R 700 /var/lib/caddy/.local/share/caddy/certificates/mycerts
   sudo chmod -R 600 /var/lib/caddy/.local/share/caddy/certificates/mycerts/birds.oldgristmill.ca/*
   ```
- Then, as you saw in the caddyserver.com link, you'll add the `tls yourcertificate.pem yourkey.pem` at the top of the site-block, ensuring your site-block has `https://birds.oldgristmill.ca` as a site-address. The way to do this for now is also through the web terminal or SSH using `nano`. Open the Caddyfile and edit it `sudo nano /etc/caddy/Caddyfile`. You should make the edits I highlight below to tell Caddy to use your certificates and serve them over port 8443:
   ```
   http://localhost http://birdnetpi.local https://birds.oldgristmill.ca:8443 { <------- You're adding ":8443" to "https://birds.oldgristmill.ca"
   tls yourcertificate.pem yourkey.pem                                                       <----- and you add this line. Everything else is the same.
   root * /home/pi/BirdSongs/Extracted
   file_server browse
   . . .
   }
   ```
- Once you've edited the Caddyfile, reload Caddy and view the logs while it handles the new certificate. Send me any errors it displays, OR you'll be online, ready to go :)
   ```
   sudo systemctl reload caddy && journalctl -fu caddy
   ```

The last command _follows_ the caddy log, so once you've seen enough log output to determine success or errors, you'll press Ctrl+C to stop that command.

My fingers are crossed for you, BUT I also want to let you know that `newif` works a tiny bit differently. It won't affect what you have going on now, but may be something you want to update to in the future, at which point you'll make similar adjustments to your new Caddyfile.

Please let me know what issues you encounter and I'll be happy to help,

My best,
Patrick
